chore: FIT-671: Add state transitions to the current application flows

bmartel opened the pull request:
This pull request introduces significant improvements to the context management and finite state machine (FSM) infrastructure in the codebase, with a focus on making context propagation more robust for asynchronous jobs, enhancing FSM reliability and configurability, and introducing new state choices for annotation drafts. The most notable changes include the addition of a `CurrentContext` class for thread-local context management, improved context capture for async jobs, a more flexible state transition registration decorator, and optimizations for FSM state transitions and cache handling.

**Context Management and Async Job Improvements:**

* Introduced the `CurrentContext` class in `core/current_request.py` to provide a unified, thread-local context management interface, including methods to set/get user, organization, and request, as well as utilities for FSM feature toggling and context clearing. All context access is now routed through this class, replacing direct thread-local usage.
* Enhanced async job handling in `core/redis.py` to automatically capture and serialize relevant context (user, organization, custom job data) and inject it into job metadata, ensuring context is preserved for background tasks. [[1]](diffhunk://#diff-2e8881789c63cd7d2fa057833d32a8429037cc6263a9b306679bb15fec088b55R85-R118) [[2]](diffhunk://#diff-2e8881789c63cd7d2fa057833d32a8429037cc6263a9b306679bb15fec088b55R163-L146)
* Updated job execution logic to cleanly separate context management for async (Redis) and sync jobs, and improved argument handling for job submission. [[1]](diffhunk://#diff-2e8881789c63cd7d2fa057833d32a8429037cc6263a9b306679bb15fec088b55L115-R153) [[2]](diffhunk://#diff-2e8881789c63cd7d2fa057833d32a8429037cc6263a9b306679bb15fec088b55R193) [[3]](diffhunk://#diff-2e8881789c63cd7d2fa057833d32a8429037cc6263a9b306679bb15fec088b55R203-R204)

**FSM Infrastructure Enhancements:**

* Added a feature flag check to enable or disable FSM logic dynamically, using the new context management utilities for user/organization scoping. FSM state retrieval now respects this flag, and logging includes organization context. [[1]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R12-R13) [[2]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R58-R64) [[3]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R90-R104) [[4]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4L109-R140)
* Improved the `transition_state` method in `fsm/state_manager.py` with optimistic concurrency control using cache-based locks, prevention of redundant same-state transitions, and correct mapping of entity fields for state model records. Organization context is now consistently set and logged. [[1]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R176) [[2]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R193) [[3]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R210-R280) [[4]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R295-R301)

**FSM State and Transition Registration:**

* Extended the `register_state_transition` decorator to support trigger metadata for transitions, allowing fine-grained control over when transitions are triggered (on create, update, or specific fields).
* Introduced a new `AnnotationDraftStateChoices` enum for draft annotation lifecycle states, supporting a more granular and independent workflow for annotation drafts.

**Other Notable Improvements:**

* Corrected annotation deletion logic to count items before deletion, ensuring accurate reporting of deleted annotations.

These changes collectively improve the reliability, observability, and flexibility of both context propagation and FSM-driven workflows across the application.

---

**Most Important Changes:**

_Context Management & Async Job Context Propagation:_
- Added `CurrentContext` for unified thread-local context management, replacing direct `_thread_locals` usage and providing methods for user, organization, and request handling.
- Async job submission now auto-captures and serializes context (user, organization, job data) for use in background jobs, ensuring proper context propagation. [[1]](diffhunk://#diff-2e8881789c63cd7d2fa057833d32a8429037cc6263a9b306679bb15fec088b55R85-R118) [[2]](diffhunk://#diff-2e8881789c63cd7d2fa057833d32a8429037cc6263a9b306679bb15fec088b55R163-L146)

_FSM Infrastructure & Transition Handling:_
- FSM state retrieval and transitions now respect a feature flag and include optimistic concurrency control, prevention of redundant transitions, and improved logging with organization context. [[1]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R58-R64) [[2]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R90-R104) [[3]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4L109-R140) [[4]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R176) [[5]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R193) [[6]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R210-R280) [[7]](diffhunk://#diff-757222088adfaa313e095fd7d461743038d7cf0963ca8ca443ba186cd0e6a9f4R295-R301)
- Enhanced `register_state_transition` decorator to support triggers on create, update, or specific fields, enabling more flexible FSM workflows.

_FSM State Choices:_
- Added `AnnotationDraftStateChoices` to support a full draft annotation lifecycle, independent from main annotation states.

_Annotation Deletion:_
- Fixed annotation deletion logic to count items before deletion for accurate reporting.

bmartel committed:
feat: FIT-587: FSM Architecture setup

bmartel committed:
include organization on base state model

bmartel committed:
refactoring core definitions

bmartel committed:
define core entity states in full

bmartel committed:
regen migrations for fsm core

bmartel committed:
regen migrations for fsm core

bmartel committed:
regen migrations for fsm core

bmartel committed:
fixed imports

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
fix imports

bmartel committed:
fix imports

bmartel committed:
fix fsm tests

bmartel committed:
fix fsm tests

bmartel committed:
fix linting errors

bmartel committed:
fix fsm concurrency tests

bmartel committed:
fix remaining fsm tests

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
removing implementation details of fsm to break up the PR

bmartel committed:
removing implementation details of fsm to break up the PR

bmartel committed:
removing implementation details of fsm to break up the PR

bmartel committed:
updating doc strings

bmartel committed:
fixing registry

bmartel committed:
fixing registry

bmartel committed:
fixing tests

bmartel committed:
updating README

bmartel committed:
Apply suggestion from @bmartel

bmartel committed:
avoiding import cycle due to types

bmartel committed:
feat: FIT-587: FSM core model setup

bmartel committed:
fix imports

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
fix imports

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
fix typing and decorator definition for state model

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
adding migrations

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
removing unused code

bmartel committed:
moving settings from extensions to settings file

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
moving the API integration FSM out of this PR

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
updating readme

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
removing unused code

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
renaming transition decorator for consistency

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
fixing tests to align with refactors and renames of transition functions

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
removing unused code

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
removing serializers from this PR, will be included in a different one

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-587-fsm-models

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
Update label_studio/fsm/state_manager.py

Co-authored-by: Marcel Canu <marcel@humansignal.com>

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
updating based on feedback about test assertions/file names

bmartel committed:
updating based on feedback about test file names

bmartel committed:
consolidate transition util get_valid_states to being the same function get_available_states

bmartel committed:
improving transition handling

bmartel committed:
improving transition handling by consolidating everything to use a single entry point

bmartel committed:
ensure write through cache sets on_commit of the transaction

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
type fixes

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
update tests to ensure coverage

bmartel committed:
fixing lint errors

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17446383898

bmartel committed:
Merge branch 'develop' into 'fb-fit-587-fsm'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17446383898

bmartel committed:
reverting poetry.lock to develop to regenerate it on CI

bmartel committed:
Merge branch 'fb-fit-587-fsm' of github.com:HumanSignal/label-studio into fb-fit-587-fsm

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17446529166

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
updating docs

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
removing unused code

bmartel committed:
applying feedback from CR

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
fixing tests

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-587-fsm

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
structure the logging so it can be made useful for metrics in grafana

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
fix tests

bmartel committed:
fix tests

bmartel committed:
fix implementation

bmartel committed:
fix implementation

bmartel committed:
Merge branch 'develop' into fb-fit-587-fsm

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17496376340

bmartel committed:
Merge branch 'develop' into 'fb-fit-587-fsm'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17496376340

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
fix initial migration

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17497319557

bmartel committed:
Merge branch 'develop' into 'fb-fit-587-fsm'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17497319557

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
refactor code to remove function imports

bmartel committed:
refactor code to remove redundant code

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
remove no longer required test

bmartel committed:
updating docs

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
fixing lint errors

bmartel committed:
fixing lint errors

bmartel committed:
Merge branch 'fb-fit-587-fsm' into fb-fit-610-fsm-models

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-610-fsm-models

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17554612979

bmartel committed:
Merge branch 'develop' into 'fb-fit-610-fsm-models'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17554612979

bmartel committed:
fixing tests for windows SQLITE

bmartel committed:
Merge branch 'fb-fit-610-fsm-models' of github.com:HumanSignal/label-studio into fb-fit-610-fsm-models

bmartel committed:
Update label_studio/fsm/models.py

Co-authored-by: Marcel Canu <marcel@humansignal.com>

bmartel committed:
Update label_studio/fsm/models.py

Co-authored-by: Marcel Canu <marcel@humansignal.com>

bmartel committed:
applying feedback to remove the additional method get_current_state and utilize get_current_state_value and get_current_state_object as the only means for retrieving current_state

bmartel committed:
removing redundant db_index=True on task fk

bmartel committed:
using factories as mentioned in PR feedback

bmartel committed:
feat: FIT-671: Tracking entity state change with transitions

bmartel committed:
feature flagging changes

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
ensure integrations are behind FF

bmartel committed:
test fixes

bmartel committed:
improve coverage

bmartel committed:
fix init of integrations to only include the correct version of integrations

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17658721436

bmartel committed:
Merge branch 'develop' into 'fb-fit-671'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17658721436

bmartel committed:
fixing deletion counts to not include cascade records

bmartel committed:
Merge branch 'fb-fit-671' of github.com:HumanSignal/label-studio into fb-fit-671

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17677626300

bmartel committed:
Merge branch 'fb-fit-671' of github.com:HumanSignal/label-studio into fb-fit-671

bmartel committed:
removing md file

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17678572623

bmartel committed:
Merge branch 'develop' into 'fb-fit-671'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17678572623

bmartel committed:
updating logging to ensure it is following the same structure everywhere

bmartel committed:
Merge branch 'fb-fit-671' of github.com:HumanSignal/label-studio into fb-fit-671

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
cleanup move to native python asserts for tests

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17808130236

bmartel committed:
Merge branch 'develop' into 'fb-fit-671'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17808130236

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
adding context cache for organization_id lookup to avoid n+1

bmartel committed:
Apply pre-commit linters

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17838626349

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17843084400

bmartel committed:
Merge branch 'develop' into 'fb-fit-671'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17843084400

bmartel committed:
fix writethrough cache concurrency

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
fixing tests

bmartel committed:
fixing tests

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
switch to a more flexible current context that can work either in requests or jobs, as well as ASGI environments

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
trying another way of sharing context

bmartel committed:
trying another solution to shared context

bmartel committed:
another try

bmartel committed:
the imports of CurrentContext have to be lazy

bmartel committed:
split job data and thread local data so we can retain the requesting user and org id for the entire chain of request and async jobs while allowing thread local data

bmartel committed:
must use safe access

bmartel committed:
wrong method name

bmartel committed:
removing unused code

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671

bmartel committed:
removing unused code

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel committed:
trying with the other method of cleanup

bmartel committed:
trying different import

bmartel committed:
simplify all the current context back to the same file

Commit:
Mocked up CurrentContext

Commit:
Unstubbing other methods

bmartel committed:
trying without job data:

bmartel committed:
clear all data from current context on request end

bmartel committed:
fsm context with worker jobs

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/18048942826)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18048942826

bmartel committed:
Merge branch 'develop' into 'fb-fit-671-thread-context'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18048942826

bmartel committed:
allow detection for async job

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/18099038603)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18099038603

bmartel committed:
Merge branch 'develop' into 'fb-fit-671-thread-context'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18099038603

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel committed:
Merge branch 'fb-fit-671-thread-context' of github.com:HumanSignal/label-studio into fb-fit-671-thread-context

bmartel committed:
unused code

bmartel committed:
use current context for org id

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/18192838961)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18192838961

bmartel committed:
Merge branch 'develop' into 'fb-fit-671-thread-context'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18192838961

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

makseq commented on label_studio/fsm/signals.py:
```diff
@@ -0,0 +1,177 @@
+    return None
+
+
+@receiver(post_save)
```
Please don't use signals, it's very hard to debug them and trace the order of their calls.

bmartel replied:
What alternatives would you suggest to keep this loosely coupled?

wesleylima replied:
You could use a base model or mixin to override the save and update methods, but I think that's not better than signals because it increases the surface area of changes, since we have to update every model definition (which may also cause some `noop` migrations to be generated) and we'll also have to patch in FSM into every `.save()` and `.update()` override that are already in place.

I think any other alternative would create _much_ tighter coupling, like adding a "service" layer between the API views and the model, which would also be a significant departure from the normal Django app architecture.

@makseq is right, signals are normally very hard to debug, but I think FSM right now has some very good logging and the data entities in FSM itself help the entire chain of execution to be very debuggable.

jombooth replied:
I think the base model idea is a pretty good one - ie, subclass `models.Model` as `HsModel` and replace all `models.Model` references with it. I don't think this would generate any migrations, and it would avoid some of the pain associated with signals, which I can attest is very real (predictable backend testing, in particular, can become a real challenge when signals are in the picture)

bmartel replied:
Done 👍

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

jombooth commented on label_studio/fsm/signals.py:
```diff
@@ -0,0 +1,177 @@
+
+
+@receiver(post_save)
+def handle_model_state_transitions(sender, instance, created, **kwargs):
```
I think it would be nice to break this up a bit, from a large if..elif..elif... wrapped in a try-except. How about:

1) implementing the error handling in a decorator
2) implementing a handler for each model as a function, decorated with said decorator
3) using a "registry" mapping each model name to its handler function
4) when we are going to do a transition, pull the function from the registry and call it. Log appropriately if no handler is registered for the model in question

This would make it easier to test each state transition handler in a decoupled fashion, as well as to gradually add new ones and increase their complexity over time

bmartel replied:
Done 👍, uses explicit modelled Transition objects instead.

jombooth commented on label_studio/fsm/integrations.py:
```diff
@@ -0,0 +1,372 @@
+
+    # Check if this is Enterprise edition via Django settings
+    if getattr(settings, 'VERSION_EDITION', 'Community') != 'Community':
+        return True
```
I recommend reusing `from core.utils.common import is_community`. We've generally framed this as community vs not so far, because technically, Starter Cloud uses the LSE codebase while not being "enterprise" per se

bmartel replied:
Refactored away from this file entirely, and the approach is using registration as the core means to extension for all aspects of FSM.

jombooth commented on label_studio/core/redis.py:
```diff
@@ -80,6 +82,40 @@ def redis_connected():
     return redis_healthcheck()
 
 
+def _is_serializable(value: Any) -> bool:
+    """Check if a value can be serialized for job context."""
+    return isinstance(value, (str, int, float, bool, list, dict, type(None)))
```
In what sense do we mean serializable here - ie, into what format? it strikes that including `list` and `dict` in here could be risky depending on the notion of serializability we mean

bmartel replied:
Great point, I was trying to cut down on accidental passing of values that were not expected, like classes etc. This was a means to ensure there were only simple values being passed, not complex objects.

jombooth commented on label_studio/fsm/state_manager.py:
```diff
@@ -190,100 +207,155 @@ def transition_state(
-                    **denormalized_fields,
-                )
+            try:
+                with transaction.atomic():
```
why is transaction.atomic() needed here? reading through, it looked like only one DB operation happens here (state_model.objects.create()). Very possible I'm missing something here

bmartel replied:
That is a great question, honestly after the multitude of iterations here, I don't see a need for it. This used to be differently handled in one incarnation I had, but since that point with different refactors and changes in approach, this does not look effective anymore.

bmartel committed:
implement base model and a declarative transition system

bmartel committed:
refactor base model implementation with declarative transitions

bmartel committed:
remove all signal related concepts

bmartel committed:
refactor and change tests

bmartel committed:
refactor all tests and fsm code to use model based approach

bmartel committed:
fixing fsm tests and improving integration coverage

bmartel committed:
improving fsm coverage

bmartel committed:
more fsm test fixes

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

jombooth commented on label_studio/projects/__init__.py:
```diff
@@ -1,2 +1,10 @@
+    from projects import transitions  # noqa: F401
+except ImportError:
+    # FSM may not be available in all configurations
+    pass
```
I think we should definitely log if we go down the exception branch - this could easily be missed in a refactor of the projects module for example. As in the example from LSE, let's log in any places where we handle an ImportError

bmartel replied:
Yea, I actually was wondering about this flow, I might revisit it as it seems like it should be part of the apps.py ready callback of a given app to avoid any loading funny business and to keep it inline with the Django lifecycle.

jombooth reviewed (approved):
LGTM once the pytests are green!

bmartel committed:
fix recursion errors caused by the transition imports, fix some of the failing tests when FSM is enabled

bmartel committed:
move fsm enablement to fixtures, default on

triklozoid commented on label_studio/core/redis.py:
```diff
@@ -122,28 +160,26 @@ def start_job_async_or_sync(job, *args, in_seconds=0, **kwargs):
         del kwargs['job_timeout']
+
     if redis:
-        # Auto-capture request_id from thread local and pass it via job meta
```
It looks like this change breaks passing `request_id` into the job.

bmartel replied:
It still does it, just using CurrentContext to handle all the thread_local related aspects. I had this solution on branch before this change for request_id, so I just merged it together. request_id is treated just as other job_data is, and is part of a dictionary of values that is set onto the job meta and restored from there on the job side.

bmartel committed:
cleanup and fix non fsm tests

bmartel committed:
cleanup and fix non fsm core tests

bmartel committed:
isolating non fsm failures by temporarily removing fsm integration tests from projects and tasks apps

bmartel committed:
removing unused code

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel committed:
fix projects and tasks apps config

bmartel committed:
fix the base FSM transitions and add tests for the workflows

bmartel committed:
fix StateManager reference

bmartel committed:
fix anno delete state transitions for tasks

bmartel committed:
fixed fsm cold start scenarios

bmartel committed:
trying a forced import to see if it affects the fsm workflow tests on CI

bmartel committed:
fixing fsm registration

bmartel committed:
fixing fsm registration

bmartel committed:
fixing fsm registration

bmartel committed:
removing fsm tests outside of fsm app tests

bmartel committed:
fix fsm test teardown/setup

bmartel committed:
restoring the lso workflows test

bmartel committed:
leave the INSTALLED_APPS to match the syntax of the surrounding

bmartel committed:
fixing core context and async job detection

bmartel committed:
fixing redis and base model changed fields handling

bmartel committed:
fixing recursion error in tests

bmartel committed:
applying PR feedback

bmartel committed:
updating base state manager handling

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel committed:
fix import registration of fsm transitions

bmartel committed:
fixing import storage sync to ensure that we use start_job_async_or_sync to make CurrentContext trace through for FSM and request_id

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

bmartel committed:
ensure tasklock uses fsm basemodel

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19040255769)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19040255769

bmartel committed:
Merge branch 'develop' into 'fb-fit-671-thread-context'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19040255769

hlomzik committed:
feat: BROS-587: Support HTML in Markdown (#8747)

Co-authored-by: hlomzik <hlomzik@users.noreply.github.com>

nick-skriabin committed:
fix: BROS-464: Vector region selection (#8636)

Co-authored-by: robot-ci-heartex <robot-ci-heartex@users.noreply.github.com>
Co-authored-by: hlomzik <hlomzik@users.noreply.github.com>
Co-authored-by: nick-skriabin <nick-skriabin@users.noreply.github.com>

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19076424295)

bmartel commented on label_studio/core/settings/base.py:
```diff
@@ -216,13 +216,14 @@
-    'projects',
-    'tasks',
+    'projects',  # Registers FSM transitions - needs fsm loaded first
+    'tasks',  # Registers FSM transitions - needs fsm loaded first
```
```suggestion
    'projects',
    'tasks', 
```

bmartel committed:
Apply suggestion from @bmartel

bmartel commented on label_studio/core/settings/base.py:
```diff
@@ -216,13 +216,14 @@
     'data_export',
     'projects',
-    'tasks',
+    'tasks', 
```
```suggestion
    'tasks',
```

bmartel committed:
Apply suggestion from @bmartel

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19076424295

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-671-thread-context

hakan458 commented on label_studio/io_storages/base_models.py:
```diff
@@ -665,10 +698,13 @@ def sync(self):
                     import_sync_background,
                     self.__class__,
                     self.id,
+                    queue_name='low',
```
can get rid of `queue = django_rq.get_queue('low')` above now to avoid confusion

bmartel replied:
It is used  by the `is_job_in_queue` function, which is why I left it.  I was thinking in another pass (another PR) we would go through and scrub all the extras and rework things accordingly to do so.

makseq replied:
It seems we can replace enqueue to start_job_async_or_sync, but take into account:
https://github.com/HumanSignal/label-studio/pull/4059/files#r1191788434

It means some checks, such as if if redis_connected(), will be called twice. Maybe something else will be affected... please double-check it.

makseq commented on label_studio/core/models.py:
```diff
@@ -9,6 +10,569 @@
 logger = logging.getLogger(__name__)
 
 
+class HsModel(models.Model):
```
why do we need to save it in /core? and not in /fsm? or does it lead to circular imports? 
I would try keeping isolate fsm maximally

makseq replied:
Also, I would name it FsmHistoryStateModel, it's much more clearer

makseq commented on label_studio/core/models.py:
```diff
@@ -9,6 +10,569 @@
+            Whatever super().save() returns
+        """
+        # Check for explicit FSM skip flag
+        skip_fsm = kwargs.pop('skip_fsm', False)
```
I would make it True by default to ensure we don't affect existing code

makseq commented on label_studio/core/redis.py:
```diff
@@ -122,28 +160,26 @@ def start_job_async_or_sync(job, *args, in_seconds=0, **kwargs):
-        except Exception:
-            # Fail silently if no request context
-            pass
+        # Async execution with Redis - wrap job for context management
```
try/except is missed here now

makseq commented on label_studio/io_storages/base_models.py:
```diff
@@ -484,6 +485,8 @@ def add_task(cls, project, maximum_annotations, max_inner_id, storage, link_obje
                 inner_id=max_inner_id,
             )
+            # Save with skip_fsm flag to bypass FSM during bulk import
+            task.save(skip_fsm=True)
```
I would keep skip_fsm=True by default and avoid modifying extra code

bmartel replied:
But then we will have to make more changes than just skipping one case, we'd have to tell the 99% of cases we want to capture to do so. I don't think that makes sense. That would mean more code

bmartel replied:
This is Feature Flagged and the system won't do anything at all if off. When on we need this to be handled ubiquitously otherwise we will miss states. This is required to be this way.

makseq commented on label_studio/io_storages/base_models.py:
```diff
@@ -640,6 +643,36 @@ def _scan_and_create_links(self, link_class):
+
+                logger.info(f'Storage sync: FSM states created for {len(task_ids)} tasks')
+            except ImportError:
+                # LSE not available (OSS), skip FSM sync
+                pass
+            except Exception as e:
+                # Don't fail storage sync if FSM sync fails
+                logger.error(f'FSM sync after storage sync failed: {e}', exc_info=True)
+
```
it's better to move to /fsm/import.py or /fsm/io_storages.py or maybe functions.py or whatever; and write here just one line that calls a function from /fsm/import.py (or whatever)

makseq commented on label_studio/tasks/models.py:
```diff
@@ -773,13 +774,65 @@ def save(self, *args, update_fields=None, **kwargs):
+
         return result
 
+    def _update_task_state_after_deletion(self, task, project):
```
I would move it to /fsm/tasks.py or /fsm/functions.py

bmartel committed:
refactor all fsm definitions to fsm app

bmartel committed:
applying PR feedback

bmartel committed:
applying PR feedback

bmartel committed:
cleanup imports

bmartel commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19137772912)
> Successfully merged:  create mode 100644 web/libs/ui/src/lib/enterprise-upgrade-overlay/index.ts

bmartel committed:
Merge branch 'develop' into 'fb-fit-671-thread-context'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19137772912
